src/main/java/io/cdap/plugin/gcp/bigquery/relational/BigQueryRelation.java

```diff
@@ -1,117 +1,409 @@
 package io.cdap.plugin.gcp.bigquery.relational;
 
+import com.google.common.annotations.VisibleForTesting;
 import io.cdap.cdap.etl.api.aggregation.DeduplicateAggregationDefinition;
 import io.cdap.cdap.etl.api.aggregation.GroupByAggregationDefinition;
 import io.cdap.cdap.etl.api.relational.Expression;
+import io.cdap.cdap.etl.api.relational.InvalidRelation;
 import io.cdap.cdap.etl.api.relational.Relation;
+import io.cdap.plugin.gcp.bigquery.sqlengine.BigQuerySQLDataset;
+import io.cdap.plugin.gcp.bigquery.sqlengine.builder.BigQueryDeduplicateSQLBuilder;
+import io.cdap.plugin.gcp.bigquery.sqlengine.builder.BigQueryGroupBySQLBuilder;
+import io.cdap.plugin.gcp.bigquery.sqlengine.builder.BigQueryNestedSelectSQLBuilder;
+import io.cdap.plugin.gcp.bigquery.sqlengine.builder.BigQuerySelectSQLBuilder;
 
-import java.util.HashMap;
+import java.util.Collection;
+import java.util.LinkedHashMap;
 import java.util.Map;
 import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
 
 /**
  * An implementation of {@link Relation} designed to operate on BigQuery.
  */
 public class BigQueryRelation implements Relation {
-  private final boolean isValid;
-  private final String validationError;
+  private static final SQLExpressionFactory factory = new SQLExpressionFactory();
 
-  private final Map<String, Expression> columns;
-  private final Expression filter;
-  private final Relation baseRelation;
   private final String datasetName;
+  private final Set<String> columns;
+  private final BigQueryRelation parent;
+  private Map<String, BigQuerySQLDataset> sourceDatasets;
+  private final Supplier<String> expressionSupplier;
```

Comment:
nit: name it `sqlExpressionSupplier` or even `sqlStatementSupplier` to highlight what's that. Otherwise the name is very generic and does not tell what's inside